snuba/consumers/strict_consumer.py

```diff
@@ -0,0 +1,141 @@
+    from the current committed offset to the end of the topic and then terminates.
+    It waits for a partition to be assigned before actually consuming. If no
+    partition is assigned within a timeout it fails.
+    It also moves the responsibility of deciding what to commit to the user
+    compared to the BatchingConsumer
+
+    This is not supposed to be used for continuously consume a topic but to
+    catch up on a topic from the beginning to the state it is at the time we start.
+    An example is the control topic when taking snapshots.
+    """
+
+    def __init__(
+        self,
+        topic: str,
+        bootstrap_servers: Sequence[str],
+        group_id: str,
+        auto_offset_reset: str,
+        partition_assignment_timeout: int,
+        on_partitions_assigned: Callable[[Consumer, Sequence[TopicPartition]], None],
+        on_partitions_revoked: Callable[[Consumer, Sequence[TopicPartition]], None],
+        on_message: Callable[[Message], CommitDecision],
+    ) -> None:
+        self.__on_partitions_assigned = on_partitions_assigned
+        self.__on_partitions_revoked = on_partitions_revoked
+        self.__on_message = on_message
+        self._assign_timeout = partition_assignment_timeout
+
+        self.__shutdown = False
+
+        consumer_config = {
+            'enable.auto.commit': False,
+            'bootstrap.servers': ','.join(bootstrap_servers),
+            'group.id': group_id,
+            'enable.partition.eof': 'true',
+            'default.topic.config': {
+                'auto.offset.reset': auto_offset_reset,
+            },
+        }
+
+        self.__consumer = self.create_consumer(consumer_config)
+
+        def on_partitions_assigned(
+            consumer: Consumer,
+            partitions: Sequence[TopicPartition],
+        ):
+            logger.info("New partitions assigned: %r", partitions)
+            self.__on_partitions_assigned(consumer, partitions)
+
+        def on_partitions_revoked(
+            consumer: Consumer,
+            partitions: Sequence[TopicPartition],
+        ):
+            logger.info("Partitions revoked: %r", partitions)
+            self.__on_partitions_revoked(consumer, partitions)
+
+        self.__consumer.subscribe(
+            [topic],
+            on_assign=on_partitions_assigned,
+            on_revoke=on_partitions_revoked,
+        )
+
+    def create_consumer(self, config) -> Consumer:
+        return Consumer(config)
+
+    def run(self) -> None:
+        watermarks = {}
+
+        message = self.__consumer.poll(timeout=self._assign_timeout)
+        if not message:
+            # Since we enabled enable.partition.eof, no message means I could not
+            # consume and it is different from the case I managed to consume and
+            # I reached the end of the partition. Thus we fail.
+            raise NoPartitionAssigned("No partition was assigned within timeout")
```

Comment:
I wonder if it might be clearer to use the assignment callback for this?

Reply:
I cannot, this condition means I cannot go ahead if a partition does not get assigned before a timeout (because I need to consume the whole topic before going ahead). If I cannot consumer because no partition gets assigned the correct behavior is failing instead of assuming there was no snapshot.